### PR TITLE
Parse HTTP proxy credentials from URL and populate settings

### PR DESCRIPTION
**Pull request by krlmlr**
## Summary
Enhanced HTTP proxy configuration to automatically extract embedded credentials from proxy URLs and populate the corresponding username/password settings. This allows users to specify proxy credentials in a single URL string rather than managing separate settings.

## Key Changes
- **Credential extraction**: Added `ApplyHTTPProxyURL()` function that parses proxy URLs and extracts `user[:password]@` userinfo components, storing them in `http_proxy_username` and `http_proxy_password` settings
- **RFC 3986 compliance**: Correctly handles userinfo parsing by only treating `@` characters in the authority component as credential separators, ignoring `@` in path/query/fragment components
- **Environment variable precedence**: Updated `HTTPProxySetting::ResetGlobal()` to check environment variables in order: `https_proxy`, `HTTPS_PROXY`, `http_proxy`, `HTTP_PROXY` (preferring lowercase to avoid CGI httpoxy ambiguity)
- **URL normalization**: Removes extracted credentials from the stored proxy URL, leaving only the scheme and authority
- **Comprehensive test coverage**: Added 246-line test file covering valid URLs, edge cases (empty userinfo, multiple `@` characters, colons in passwords), and malformed URLs

## Implementation Details
- Credential settings are only modified when non-empty userinfo is found in the URL
- Empty userinfo (`@host`) or userinfo in query/fragment components are handled gracefully without affecting existing credentials
- Explicit empty password (`user:@host`) clears the password while preserving the username
- The implementation is robust to malformed URLs, leaving them unchanged rather than failing
- Updated setting descriptions to document the new credential extraction behavior

https://claude.ai/code/session_01GC39M53ozHzknQwf3M68BJ

### CODE REVIEW

File: src/main/settings/custom_settings.cpp [hunk 1]
```diff
@@ -1265,12 +1265,60 @@ Value HTTPLoggingOutputSetting::GetSetting(const ClientContext &context) {
...
 // HTTP Proxy
 //===----------------------------------------------------------------------===//
+namespace {
+
+// Assign a proxy URL to the `http_proxy` option, splitting any embedded
+// `user[:password]@` userinfo into the matching `http_proxy_username` /
+// `http_proxy_password` settings. The credential settings are only touched
+// when a non-empty userinfo is found.
+void ApplyHTTPProxyURL(DBConfig &config, string proxy) {
+	idx_t userinfo_start = 0;
```
Comment (krlmlr) [inline src/main/settings/custom_settings.cpp]: The 0 case is not very well defined. Stop parsing the URL and just set the proxy variable and nothing else in this case. Use early return, move the declaration past the `if` block. Add a test.

File: src/main/settings/custom_settings.cpp [hunk 2]
```diff
@@ -1265,12 +1265,60 @@ Value HTTPLoggingOutputSetting::GetSetting(const ClientContext &context) {
...
+	auto scheme_end = proxy.find("://");
+	if (scheme_end != string::npos) {
+		userinfo_start = scheme_end + 3;
+	}
+	// Userinfo (if any) lives in the authority component, which ends at the
+	// first '/', '?' or '#' character (RFC 3986 §3.2); a later '@' belongs to
+	// the path / query / fragment and must not be treated as a credential.
+	auto authority_end = proxy.find_first_of("/?#", userinfo_start);
+	auto at_pos = proxy.find('@', userinfo_start);
+	if (at_pos != string::npos && (authority_end == string::npos || at_pos < authority_end)) {
```
Comment (krlmlr) [inline src/main/settings/custom_settings.cpp]: Early exit (shortcut) here too?

File: src/main/settings/custom_settings.cpp [hunk 3]
```diff
@@ -1265,12 +1265,64 @@ Value HTTPLoggingOutputSetting::GetSetting(const ClientContext &context) {
...
+	idx_t userinfo_start = scheme_end + 3;
+	// Userinfo (if any) lives in the authority component, which ends at the
+	// first '/', '?' or '#' character (RFC 3986 §3.2); a later '@' belongs to
+	// the path / query / fragment and must not be treated as a credential.
+	auto authority_end = proxy.find_first_of("/?#", userinfo_start);
+	auto at_pos = proxy.find('@', userinfo_start);
+	if (at_pos == string::npos || (authority_end != string::npos && at_pos > authority_end)) {
+		config.options.http_proxy = std::move(proxy);
+		return;
+	}
```
Comment (krlmlr) [inline src/main/settings/custom_settings.cpp]: Can do two early exits now and compute `authority_end` only if needed?

File: src/main/settings/custom_settings.cpp [hunk 4]
```diff
@@ -1265,12 +1265,64 @@ Value HTTPLoggingOutputSetting::GetSetting(const ClientContext &context) {
...
+	auto userinfo = proxy.substr(userinfo_start, at_pos - userinfo_start);
+	// Drop `userinfo@` from the URL whether or not credentials were extracted,
+	// so the remainder is the bare proxy authority.
+	proxy.erase(userinfo_start, at_pos - userinfo_start + 1);
+	if (!userinfo.empty()) {
```
Comment (krlmlr) [inline src/main/settings/custom_settings.cpp]: How can `userinfo` be empty here? Add a comment or drop the check.

Comment (krlmlr) [reply src/main/settings/custom_settings.cpp]: Shortcut here too.

**Review by millonbigs-cpu — APPROVED** [review]

**Review by millonbigs-cpu — APPROVED** [review]

**Review by millonbigs-cpu — APPROVED** [review]
